feat: conditional coloring for big number chart

gbusch opened the pull request:
### SUMMARY
As suggested here: #21820
allow for conditional formatting (coloring) of Big Number charts.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<img width="1261" alt="Screenshot 2023-02-12 at 15 58 36" src="https://user-images.githubusercontent.com/16560193/218318705-20c14e5e-a091-4d57-a02d-2cce9324b155.png">

### TESTING INSTRUCTIONS
* select a "Big Number" chart
* add "conditional formatting" rules in the "Customize" tab

### ADDITIONAL INFORMATION
re-uses existing "ConditionalFormattingControl" from chart table and pivot table

Commit:
conditional coloring for big number chart

Commit:
solve lint issues

rusackas commented:
/testenv up

rusackas commented on superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts:
```diff
@@ -89,6 +90,45 @@ export default {
+              type: 'ConditionalFormattingControl',
+              renderTrigger: true,
+              label: t('Conditional Formatting'),
+              description: t('Apply conditional color formatting to metrics'),
```
A bit pedantic, but should it just be "metric" in this case?

gbusch replied:
true, done

rusackas commented on superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts:
```diff
@@ -188,6 +188,22 @@ describe('getColorFunction()', () => {
+      false,
+    );
+    expect(colorFunction(50)).toEqual('#FF0000');
+    expect(colorFunction(100)).toEqual('#FF0000');
```
Would it be worth adding an assertion in the middle of the range (e.g. 75) for good measure?

I'm also not sure if there's another test that asserts against values outside the range (e.g. 25 or 125 in this case). If not, that might be good to add in this test or a new one.

gbusch replied:
added tests, please resolve it ok

rusackas commented:
Code looks good, and it seems to work well on the ephemeral environment. The only question I have is whether it's possible/reasonable to make the colors full opacity rather than the sort of pastel colors that they are now. I'm a little worried about contrast/legibility in particular.

justin-tomlinson commented:
Very useful improvement. It would be great if this was available across the Big Number and Big Number with trend line. Some additional improvements to functionality could be:

- Make the colour scheme be more configurable by using hex codes or RGB values
- Making the colour selection gradient optional. sometimes you want it to be a set colour when it meets the rule rather than be some gradient of the colour. I.e If measure >0.5 then Red, not a gradient of red.

This probably applies to the conditional formatting on the table chart too. -

rusackas commented:
@gbusch curious about your input on any of the above comments... would love to see this PR get through! :)

Commit:
add more colors

gbusch commented on superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts:
```diff
@@ -89,6 +90,45 @@ export default {
+                  : explore?.datasource?.columns ?? {};
+                const { colnames, coltypes } =
+                  chart?.queriesResponse?.[0] ?? {};
+                const numericColumns =
```
might make sense but would prefer not to do it here

gbusch commented:
Sorry, was busy and then almost forgot about it...

Colors are defined in this file: 
superset-frontend/src/explore/components/controls/ConditionalFormattingControl/FormattingPopoverContent.tsx

I renamed the colors to make it clearer that they refer to the theme colors and added possibility for darker colors (from theme). Does this make sense?

gbusch commented:
@rusackas What do you think?
Pipeline is red... Don't understand why, as the changes are rather small.

kgabryje commented:
> Pipeline is red... Don't understand why, as the changes are rather small.

The E2E tests are sometimes flaky... I restarted the CI, hopefully it passes this time

mtrentz commented:
Big Numbers doesn't get color formatted when its value is zero. Anyone else experience this?

If I set a rule for "error when < 5" and the value is 0, nothing happens.

If the value is 0.0001 then it colors.

rusackas commented:
> Big Numbers doesn't get color formatted when its value is zero. Anyone else experience this?
> 
> If I set a rule for "error when < 5" and the value is 0, nothing happens.
> 
> If the value is 0.0001 then it colors.

Would you be willing to open a PR or an Issue about this? We haven't heard much outcry, and would love any help you can provide to set it on a course toward resolution.

mtrentz commented:
@rusackas opened this one #28038
